cpp/src/parquet/level_conversion_inc.h

```diff
@@ -0,0 +1,146 @@
+// under the License.
+#pragma once
+
+#include "parquet/level_conversion.h"
+
+#include <algorithm>
+#include <limits>
+#if defined(ARROW_HAVE_BMI2)
+#if defined(_MSC_VER)
+#include <immintrin.h>
+#else
+#include <x86intrin.h>
+#endif  // _MSC_VER
+#endif  // ARROW_HAVE_BMI2
+
+#include "arrow/util/bit_run_reader.h"
+#include "arrow/util/bit_util.h"
+#include "arrow/util/logging.h"
+#include "parquet/exception.h"
+#include "parquet/level_comparison.h"
+
+namespace parquet {
+namespace internal {
+namespace BMI_RUNTIME_VERSION {
+
+using ::arrow::internal::BitRun;
+using ::arrow::internal::BitRunReader;
+
+/// Algorithm to simulate pext using BitRunReader for cases where all bits
+/// not set or set.
+uint64_t RunBasedExtractMixed(uint64_t bitmap, uint64_t select_bitmap) {
+  bitmap = arrow::BitUtil::FromLittleEndian(bitmap);
+  uint64_t new_bitmap = 0;
+  ::arrow::internal::BitRunReader selection(reinterpret_cast<uint8_t*>(&select_bitmap),
+                                            /*start_offset=*/0, /*length=*/64);
+  ::arrow::internal::BitRun run = selection.NextRun();
+  int64_t selected_bits = 0;
+  while (run.length != 0) {
+    if (run.set) {
+      new_bitmap |= (bitmap & ::arrow::BitUtil::LeastSignficantBitMask(run.length))
+                    << selected_bits;
+      selected_bits += run.length;
+    }
+    bitmap = bitmap >> run.length;
+    run = selection.NextRun();
+  }
+  return arrow::BitUtil::ToLittleEndian(new_bitmap);
+}
+
+inline uint64_t RunBasedExtractImpl(uint64_t bitmap, uint64_t select_bitmap) {
+  /// These checks should be inline and are likely to be common cases.
+  if (select_bitmap == ~uint64_t{0}) {
+    return bitmap;
+  } else if (select_bitmap == 0) {
+    return 0;
+  }
+  /// Fallback to the slow method.
+  return RunBasedExtractMixed(bitmap, select_bitmap);
+}
+
+inline uint64_t ExtractBits(uint64_t bitmap, uint64_t select_bitmap) {
+#if defined(ARROW_HAVE_BMI2) && !defined(__MINGW32__)
+  return _pext_u64(bitmap, select_bitmap);
+#else
+  return RunBasedExtractImpl(bitmap, select_bitmap);
+#endif
+}
+
+template <bool has_repeated_parent>
+int64_t DefinitionLevelsBatchToBitmap(const int16_t* def_levels, const int64_t batch_size,
+                                      int64_t upper_bound_remaining, LevelInfo level_info,
+                                      ::arrow::internal::FirstTimeBitmapWriter* writer) {
+  // Greater than level_info.def_level - 1 implies >= the def_level
+  uint64_t defined_bitmap =
+      internal::GreaterThanBitmap(def_levels, batch_size, level_info.def_level - 1);
+
+  DCHECK_LE(batch_size, 64);
+  if (has_repeated_parent) {
+    // Greater than level_info.repeated_ancestor_def_level - 1 implies >= the
+    // repeated_ancenstor_def_level
+    uint64_t present_bitmap = internal::GreaterThanBitmap(
+        def_levels, batch_size, level_info.repeated_ancestor_def_level - 1);
+    uint64_t selected_bits = ExtractBits(defined_bitmap, present_bitmap);
```

Comment:
FTR, I think that if BMI isn't available, you can still use a batch size of 5 or 6 bits and use a fast lookup table for ExtractBits (rather than the probably slow emulation code).

Reply:
Yeah, we can think about that for another PR anyway. Will try to run benchmarks.